File: src/main/java/com/github/nsc/de/shake/lexer/characterinput/characterinputstream/SourceCharacterInputStream.java
package com.github.nsc.de.shake.lexer.characterinput.characterinputstream;


import com.github.nsc.de.shake.lexer.characterinput.charactersource.CharacterSource;
import com.github.nsc.de.shake.lexer.characterinput.position.PositionMaker;

import java.io.File;
import java.io.IOException;

/**
 * An implementation of {@link CharacterInputStream} using just a string as argument
 *
 * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
 */
public class SourceCharacterInputStream implements CharacterInputStream {

    /**
     * The source (mostly file) of the {@link SourceCharacterInputStream}
     */
    private final CharacterSource source;

    /**
     * The actual position of the {@link SourceCharacterInputStream}
     */
    private final PositionMaker position;


    /**
     * Constructor for {@link SourceCharacterInputStream} with given position
     *
     * @param content the characters
     * @param source the source of the characters
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     */
    public SourceCharacterInputStream( String source, char[] content) {
        this(CharacterSource.from(content, source));
    }


    /**
     * Constructor for {@link SourceCharacterInputStream} with given position
     *
     * @param content the characters
     * @param source the source of the characters
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     */
    public SourceCharacterInputStream(String source, String content) {
        this(CharacterSource.from(content, source));
    }


    /**
     * Constructor for {@link SourceCharacterInputStream} with given position
     *
     * @param file the file source
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     */
    public SourceCharacterInputStream(File file) throws IOException {
        this(CharacterSource.from(file, '<' + file.getPath() + '>'));
    }


    /**
     * Constructor for {@link SourceCharacterInputStream} with given position
     *
     * @param source the source of the characters
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     */
    public SourceCharacterInputStream(CharacterSource source) {
        // Set fields
        this.source = source;
        this.position = new PositionMaker(source);
    }


    /**
     * Constructor for {@link SourceCharacterInputStream} with given position
     *
     * @param position the starting position
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     */
    public SourceCharacterInputStream(PositionMaker position) {
        // Set fields
        this.source = position.getSource();
        this.position = position;
    }


    /**
     * Returns the source (mostly file) of the {@link CharacterInputStream}
     *
     * @return the source (mostly file) of the {@link CharacterInputStream}
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     *
     * @see CharacterInputStream#getSource()
     * @see SourceCharacterInputStream#getContent()
     * @see SourceCharacterInputStream#getPosition()
     */
    @Override
    public CharacterSource getSource() {
        // Return the source of the StringCharacterInputStream
        return source;
    }


    /**
     * Returns the chars of the {@link CharacterInputStream}
     *
     * @return the chars of the {@link CharacterInputStream}
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     *
     * @see CharacterInputStream#getContent()
     * @see SourceCharacterInputStream#getSource()
     * @see SourceCharacterInputStream#getPosition()
     */
    @Override
    public char[] getContent() {
        // Return the content of the StringCharacterInputStream
        return source.getAll();
    }


    /**
     * Returns the actual position of the {@link CharacterInputStream}
     *
     * @return the actual position of the {@link CharacterInputStream}
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     *
     * @see CharacterInputStream#getPosition()
     * @see SourceCharacterInputStream#getContent()
     * @see SourceCharacterInputStream#getSource()
     */
    @Override
    public int getPosition() {
        // Return the actual position-index of the StringCharacterInputStream
        return this.position.getIndex();
    }

    /**
     * Returns the actual position-maker of the {@link CharacterInputStream}
     *
     * @return the actual position-maker of the {@link CharacterInputStream}
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     */
    @Override
    public PositionMaker getPositionMaker() {
        // return the position of the StringCharacterInputStream
        return this.position;
    }


    /**
     * Checks if the {@link CharacterInputStream} has a next character left
     *
     * @return if the {@link CharacterInputStream} has a next character
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     *
     * @see CharacterInputStream#hasNext()
     * @see SourceCharacterInputStream#has(int number)
     */
    @Override
    public boolean hasNext() {
        // We could also use has(1) here, but for performance reasons that should be better
        return this.position.getIndex() + 1 < this.getSource().getLength();
    }


    /**
     * Checks if the {@link CharacterInputStream} has a given number of characters left
     *
     * @param number the num of characters to check
     * @return if the {@link CharacterInputStream} has a given number of characters left
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     *
     * @see CharacterInputStream#has(int number)
     * @see SourceCharacterInputStream#hasNext()
     */
    @Override
    public boolean has(int number) {
        // throw an error, if the given number is smaller than 1
        if(number < 1) throw new Error("The given number must be 1 or bigger");
        return this.position.getIndex() + number < this.getSource().getLength();
    }


    /**
     * Returns the next character and continues to the next token
     *
     * @return the next character
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     *
     * @see CharacterInputStream#next()
     */
    @Override
    public char next() {
        // Skip and return the actual character
        skip();
        return actual();
    }


    /**
     * Skips a given number of characters
     *
     * @param number the number of characters to skip
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     *
     * @see CharacterInputStream#skip(int number)
     * @see SourceCharacterInputStream#skip()
     */
    @Override
    public void skip(int number) {
        // Skip as many times, as required
        for(int i = 0; i < number; i++) skip();
    }


    /**
     * Skips the next character
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     *
     * @see CharacterInputStream#skip()
     * @see SourceCharacterInputStream#skip(int number)
     */
    @Override
    public void skip() {
        // if the actual position is a line-separator go to next line, if not then to next column
        if(this.peek() == '\n') this.position.nextLine();
        else this.position.nextColumn();
    }


    /**
     * Returns the actual character (the same as returned by {@link #next()} when used before)
     *
     * @return the actual character
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     *
     * @see CharacterInputStream#actual()
     */
    @Override
    public char actual() {
        // return the character at the actual position
        return this.getContent()[this.position.getIndex()];
    }


    /**
     * Gives back the next character without skipping
     *
     * @return the next character
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     *
     * @see CharacterInputStream#peek()
     * @see SourceCharacterInputStream#peek(int from, int to)
     * @see SourceCharacterInputStream#peek(int num)
     */
    @Override
    public char peek() {
        // we could also use peek(1) here, but for performance reasons a direct implementation is better
        // throw an error if the StringCharacterInputStream has not enough tokens left
        if(!this.hasNext()) throw new Error("Not enough characters left");

        // return the content at the required position
        return this.getContent()[this.position.getIndex() + 1];
    }


    /**
     * Gives back the next character without skipping
     *
     * @return the next character
     *
     * @param num the position to get
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     *
     * @see CharacterInputStream#peek(int num)
     * @see SourceCharacterInputStream#peek(int from, int to)
     * @see SourceCharacterInputStream#peek()
     */
    @Override
    public char peek(int num) {
        // throw an error if the StringCharacterInputStream has not enough tokens left
        // (this will also automatically throw an error, if a number that is smaller than 1 is given as input, because
        // of the implementation in the has method)
        if(!this.has(num)) throw new Error("Not enough characters left");

        // return the content at the required position
        return this.getContent()[this.position.getIndex() + num];
    }


    /**
     * Gives back a part of the {@link CharacterInputStream} as string
     * (relative to the actual position)
     *
     * @param from the starting position of the string to get
     * @param to the end position of the string to get
     * @return the character at the requested position
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     *
     * @see CharacterInputStream#peek(int from, int to)
     * @see SourceCharacterInputStream#peek(int num)
     * @see SourceCharacterInputStream#peek()
     */
    @Override
    public String peek(int from, int to) {
        if(from < 0) throw new Error("Peek argument must not be smaller than 0");
        if(to <= from) throw new Error("To-argument must be bigger than from-argument");
        //if(!this.has(from)) throw new Error("Not enough characters left");

        return this.position.getIndex() + from < this.getSource().getLength() && this.position.getIndex() + to < this.getSource().getLength() ?
                new String(this.getContent()).substring(this.position.getIndex() + from, this.position.getIndex() + to + 1) : "";
    }
}

File: util/parseutils/src/commonMain/kotlin/io/github/shakelang/parseutils/characters/streaming/SourceCharacterInputStream.kt
package io.github.shakelang.parseutils.characters.streaming

import io.github.shakelang.parseutils.characters.position.PositionMaker
import io.github.shakelang.parseutils.characters.source.CharacterSource
import io.github.shakelang.parseutils.characters.source.CharacterSource.Companion.from

/**
 * An implementation of [CharacterInputStream] using just a string as argument
 *
 * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
 */
@Suppress("unused")
class SourceCharacterInputStream(

    /**
     * The source (mostly file) of the [SourceCharacterInputStream]
     */
    override val source: CharacterSource

) : CharacterInputStream {

    /**
     * The actual position of the [SourceCharacterInputStream]
     */
    override val positionMaker: PositionMaker = PositionMaker(source)

    /**
     * Returns the actual position of the [CharacterInputStream]
     *
     * @return the actual position of the [CharacterInputStream]
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     *
     * @see CharacterInputStream.position
     * @see SourceCharacterInputStream.content
     * @see SourceCharacterInputStream.source
     */
    override val position: Int
        get() = positionMaker.index

    /**
     * Returns the chars of the [CharacterInputStream]
     *
     * @return the chars of the [CharacterInputStream]
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     *
     * @see CharacterInputStream.content
     * @see SourceCharacterInputStream.source
     * @see SourceCharacterInputStream.position
     */
    override val content: CharArray
        get() = source.all

    /**
     * Constructor for [SourceCharacterInputStream] with given position
     *
     * @param content the characters
     * @param source the source of the characters
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     */
    constructor(source: String, content: CharArray) : this(from(content, source))

    /**
     * Constructor for [SourceCharacterInputStream] with given position
     *
     * @param content the characters
     * @param source the source of the characters
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     */
    constructor(source: String, content: String) : this(from(content, source))

    /**
     * Checks if the [CharacterInputStream] has a next character left
     *
     * @return if the [CharacterInputStream] has a next character
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     *
     * @see CharacterInputStream.hasNext
     * @see SourceCharacterInputStream.has
     */
    override fun hasNext(): Boolean {
        // We could also use has(1) here, but for performance reasons that should be better
        return positionMaker.index + 1 < source.length
    }

    /**
     * Checks if the [CharacterInputStream] has a given number of characters left
     *
     * @param number the num of characters to check
     * @return if the [CharacterInputStream] has a given number of characters left
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     *
     * @see CharacterInputStream.has
     * @see SourceCharacterInputStream.hasNext
     */
    override fun has(number: Int): Boolean {
        // throw an error, if the given number is smaller than 1
        if (number < 1) throw Error("The given number must be 1 or bigger")
        return positionMaker.index + number < source.length
    }

    /**
     * Returns the next character and continues to the next token
     *
     * @return the next character
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     *
     * @see CharacterInputStream.next
     */
    override fun next(): Char {
        // Skip and return the actual character
        skip()
        return actual()
    }

    /**
     * Skips a given number of characters
     *
     * @param number the number of characters to skip
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     *
     * @see CharacterInputStream.skip
     * @see SourceCharacterInputStream.skip
     */
    override fun skip(number: Int) {
        // Skip as many times, as required
        for (i in 0 until number) skip()
    }

    /**
     * Skips the next character
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     *
     * @see CharacterInputStream.skip
     * @see SourceCharacterInputStream.skip
     */
    override fun skip() {
        // if the actual position is a line-separator go to next line, if not then to next column
        if (this.peek() == '\n') positionMaker.nextLine() else positionMaker.nextColumn()
    }

    /**
     * Returns the actual character (the same as returned by [.next] when used before)
     *
     * @return the actual character
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     *
     * @see CharacterInputStream.actual
     */
    override fun actual(): Char {
        // return the character at the actual position
        return this.content[position]
    }

    /**
     * Gives back the next character without skipping
     *
     * @return the next character
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     *
     * @see CharacterInputStream.peek
     * @see SourceCharacterInputStream.peek
     * @see SourceCharacterInputStream.peek
     */
    override fun peek(): Char {
        // we could also use peek(1) here, but for performance reasons a direct implementation is better
        // throw an error if the StringCharacterInputStream has not enough tokens left
        if (!hasNext()) throw Error("Not enough characters left")

        // return the content at the required position
        return this.content[position + 1]
    }

    /**
     * Gives back the next character without skipping
     *
     * @return the next character
     *
     * @param num the position to get
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     *
     * @see CharacterInputStream.peek
     * @see SourceCharacterInputStream.peek
     * @see SourceCharacterInputStream.peek
     */
    override fun peek(num: Int): Char {
        // throw an error if the StringCharacterInputStream has not enough tokens left
        // (this will also automatically throw an error, if a number that is smaller than 1 is given as input, because
        // of the implementation in the has method)
        if (!has(num)) throw Error("Not enough characters left")

        // return the content at the required position
        return this.content[position + num]
    }

    /**
     * Gives back a part of the [CharacterInputStream] as string
     * (relative to the actual position)
     *
     * @param from the starting position of the string to get
     * @param to the end position of the string to get
     * @return the character at the requested position
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     *
     * @see CharacterInputStream.peek
     * @see SourceCharacterInputStream.peek
     * @see SourceCharacterInputStream.peek
     */
    override fun peek(from: Int, to: Int): String {
        if (from < 0) throw Error("Peek argument must not be smaller than 0")
        if (to <= from) throw Error("To-argument must be bigger than from-argument")
        if(!this.has(to)) throw Error("Not enough characters left");
        return this.content.copyOfRange(position + from, position + to + 1).concatToString()
    }
}